File: app/src/main/java/com/niceplaces/niceplaces/activities/ExploreActivity.java
package com.niceplaces.niceplaces.activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.widget.AdapterView;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.niceplaces.niceplaces.Const;
import com.niceplaces.niceplaces.R;
import com.niceplaces.niceplaces.adapters.ListsAdapter;
import com.niceplaces.niceplaces.adapters.RegionsAdapter;
import com.niceplaces.niceplaces.controllers.AlertController;
import com.niceplaces.niceplaces.dao.DaoLists;
import com.niceplaces.niceplaces.dao.DaoPlaces;
import com.niceplaces.niceplaces.dao.DaoRegions;
import com.niceplaces.niceplaces.models.Place;
import com.niceplaces.niceplaces.utils.AppUtils;
import com.niceplaces.niceplaces.utils.ImageUtils;
import com.niceplaces.niceplaces.utils.MyRunnable;
import com.niceplaces.niceplaces.utils.NonScrollListView;

public class ExploreActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final Context context = this;
        setContentView(R.layout.activity_explore);
        getSupportActionBar().hide();
        final ExploreActivity thisActivity = this;
        FirebaseAnalytics.getInstance(this);
        final LinearLayout placeOfDayLayout = findViewById(R.id.layout_place_of_day);
        final NonScrollListView listViewHighlights = findViewById(R.id.listview_highlights);
        final NonScrollListView listViewAreas = findViewById(R.id.listview_areas);
        final AlertController alertPlaceOfDay = new AlertController(this, R.id.layout_loading_place_of_day);
        final AlertController alertLists = new AlertController(this, R.id.layout_loading_lists);
        final AlertController alertRegions = new AlertController(this, R.id.layout_loading_regions);
        DaoPlaces daoPlaces = new DaoPlaces(this);
        daoPlaces.getPlaceOfTheDay(new MyRunnable() {
            @Override
            public void run() {
                ImageView imageView = findViewById(R.id.place_of_day_image);
                TextView textViewName = findViewById(R.id.place_of_day_name);
                final TextView textViewArea = findViewById(R.id.place_of_day_area);
                final TextView textViewDesc = findViewById(R.id.place_of_day_desc);
                final Place place = getPlace();
                ImageUtils.setImageViewWithGlide(context, place.mImage, imageView);
                textViewName.setText(place.mName);
                textViewArea.setText(place.mArea + ", " + place.mRegion);
                textViewDesc.setText(place.mDescription.substring(0, 150) + "...");
                placeOfDayLayout.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        Intent intent = new Intent(thisActivity, PlaceDetailsActivity.class);
                        intent.putExtra(Const.PLACE_ID, place.getID());
                        thisActivity.startActivity(intent);
                    }
                });
                alertPlaceOfDay.loadingSuccess();
            }
        }, new Runnable() {
            @Override
            public void run() {
                alertPlaceOfDay.loadingError();
            }
        });
        DaoLists daoLists = new DaoLists(this);
        daoLists.getAll(new MyRunnable() {
            @Override
            public void run() {
                ListsAdapter adapter = new ListsAdapter(thisActivity, R.id.listview_areas, getLists());
                listViewHighlights.setAdapter(adapter);
                listViewHighlights.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                        Intent intent = new Intent(thisActivity, AreaIntroActivity.class);
                        intent.putExtra(Const.LIST_ID, getLists().get(i).getID());
                        intent.putExtra("LIST_NAME", getLists().get(i).getName());
                        intent.putExtra("LIST_DESC", getLists().get(i).getDescription());
                        thisActivity.startActivity(intent);
                    }
                });
                alertLists.loadingSuccess();
            }
        }, new Runnable() {
            @Override
            public void run() {
                alertLists.loadingError();
            }
        });
        DaoRegions daoRegions = new DaoRegions(this);
        daoRegions.getRegions(new MyRunnable() {
            @Override
            public void run() {
                RegionsAdapter adapter = new RegionsAdapter(thisActivity, R.id.listview_areas, getRegions());
                listViewAreas.setAdapter(adapter);
                listViewAreas.setOnItemClickListener(new AdapterView.OnItemClickListener() {
                    @Override
                    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                        Intent intent = new Intent(thisActivity, AreasListActivity.class);
                        intent.putExtra(Const.REGION_ID, getRegions().get(i).getID());
                        intent.putExtra("REGION_NAME", getRegions().get(i).getName());
                        thisActivity.startActivity(intent);
                    }
                });
                alertRegions.loadingSuccess();
            }
        }, new Runnable() {
            @Override
            public void run() {
                alertRegions.loadingError();
            }
        });
        final ScrollView scrollView = findViewById(R.id.scrollView);
        scrollView.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                final ViewTreeObserver.OnGlobalLayoutListener listener = this;
                scrollView.post(new Runnable() {
                    public void run() {
                        scrollView.scrollTo(0, 0);
                        scrollView.getViewTreeObserver().removeOnGlobalLayoutListener(listener);
                    }
                });
            }
        });
    }
}

File: app/src/main/java/com/niceplaces/niceplaces/activities/ExploreActivity.kt
package com.niceplaces.niceplaces.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.isVisible
import androidx.core.widget.addTextChangedListener
import androidx.core.widget.doAfterTextChanged
import com.google.firebase.analytics.FirebaseAnalytics
import com.niceplaces.niceplaces.Const
import com.niceplaces.niceplaces.R
import com.niceplaces.niceplaces.adapters.LatestPlacesAdapter
import com.niceplaces.niceplaces.adapters.ListsAdapter
import com.niceplaces.niceplaces.adapters.RegionsAdapter
import com.niceplaces.niceplaces.controllers.AlertController
import com.niceplaces.niceplaces.dao.DaoLists
import com.niceplaces.niceplaces.dao.DaoPlaces
import com.niceplaces.niceplaces.dao.DaoRegions
import com.niceplaces.niceplaces.utils.ImageUtils
import com.niceplaces.niceplaces.utils.MyRunnable
import com.niceplaces.niceplaces.utils.NonScrollListView
import java.util.*

class ExploreActivity : AppCompatActivity() {

    private lateinit var mListViewSearch: ListView
    private lateinit var mEditTextSearch: EditText

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val context: Context = this
        setContentView(R.layout.activity_explore)
        supportActionBar!!.hide()
        val thisActivity = this
        FirebaseAnalytics.getInstance(this)
        mEditTextSearch = findViewById(R.id.edittext_search)
        mListViewSearch = findViewById(R.id.listview_search_explore)
        val placeOfDayLayout = findViewById<LinearLayout>(R.id.layout_place_of_day)
        val listViewHighlights = findViewById<NonScrollListView>(R.id.listview_highlights)
        val listViewAreas = findViewById<NonScrollListView>(R.id.listview_areas)
        val alertPlaceOfDay = AlertController(this, R.id.layout_loading_place_of_day)
        val alertLists = AlertController(this, R.id.layout_loading_lists)
        val alertRegions = AlertController(this, R.id.layout_loading_regions)
        val daoPlaces = DaoPlaces(this)
        mEditTextSearch.onFocusChangeListener = View.OnFocusChangeListener { view, b ->
            if (b){
                mListViewSearch.visibility = View.VISIBLE
            } else {
                mListViewSearch.visibility = View.GONE
            }
        }
        mEditTextSearch.doAfterTextChanged { text ->
            daoPlaces.getSearchResults(text.toString().toLowerCase(Locale.ROOT).replace(" ", "+"),
                    object: MyRunnable() {
                override fun run() {
                    val adapter = LatestPlacesAdapter(thisActivity, R.id.listview_latest_places, places)
                    mListViewSearch.adapter = adapter
                    mListViewSearch.onItemClickListener = OnItemClickListener { adapterView, view, i, l ->
                        val intent = Intent(thisActivity, PlaceDetailsActivity::class.java)
                        intent.putExtra(Const.PLACE_ID, places?.get(i)?.iD)
                        thisActivity.startActivity(intent)
                    }
                }
            }, Runnable { alertPlaceOfDay.loadingError() })
        }
        daoPlaces.getPlaceOfTheDay(object : MyRunnable() {
            override fun run() {
                val imageView = findViewById<ImageView>(R.id.place_of_day_image)
                val textViewName = findViewById<TextView>(R.id.place_of_day_name)
                val textViewArea = findViewById<TextView>(R.id.place_of_day_area)
                val textViewDesc = findViewById<TextView>(R.id.place_of_day_desc)
                val place = place
                if (place != null) {
                    ImageUtils.setImageViewWithGlide(context, place.mImage, imageView)
                    textViewName.text = place.mName
                    textViewArea.text = place.mArea + ", " + place.mRegion
                }
                var desc = place?.mDescription
                if (desc != null) {
                    if (desc.length > 150) {
                        desc = desc.substring(0, 150) + "..."
                    }
                }
                textViewDesc.text = desc
                placeOfDayLayout.setOnClickListener {
                    val intent = Intent(thisActivity, PlaceDetailsActivity::class.java)
                    if (place != null) {
                        intent.putExtra(Const.PLACE_ID, place.iD)
                    }
                    thisActivity.startActivity(intent)
                }
                alertPlaceOfDay.loadingSuccess()
            }
        }, Runnable { alertPlaceOfDay.loadingError() })
        val daoLists = DaoLists(this)
        daoLists.getAll(object : MyRunnable() {
            override fun run() {
                val adapter = ListsAdapter(thisActivity, R.id.listview_areas, lists)
                listViewHighlights.adapter = adapter
                listViewHighlights.onItemClickListener = OnItemClickListener { adapterView, view, i, l ->
                    val intent = Intent(thisActivity, AreaIntroActivity::class.java)
                    intent.putExtra(Const.LIST_ID, lists?.get(i)?.iD)
                    intent.putExtra("LIST_NAME", lists?.get(i)?.name)
                    intent.putExtra("LIST_DESC", lists?.get(i)?.description)
                    thisActivity.startActivity(intent)
                }
                alertLists.loadingSuccess()
            }
        }, Runnable { alertLists.loadingError() })
        val daoRegions = DaoRegions(this)
        daoRegions.getRegions(object : MyRunnable() {
            override fun run() {
                val adapter = RegionsAdapter(thisActivity, R.id.listview_areas, regions)
                listViewAreas.adapter = adapter
                listViewAreas.onItemClickListener = OnItemClickListener { adapterView, view, i, l ->
                    val intent = Intent(thisActivity, AreasListActivity::class.java)
                    intent.putExtra(Const.REGION_ID, regions?.get(i)?.iD)
                    intent.putExtra("REGION_NAME", regions?.get(i)?.name)
                    thisActivity.startActivity(intent)
                }
                alertRegions.loadingSuccess()
            }
        }, Runnable { alertRegions.loadingError() })
        val scrollView = findViewById<ScrollView>(R.id.scrollView)
        scrollView.viewTreeObserver.addOnGlobalLayoutListener(object : OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                val listener: OnGlobalLayoutListener = this
                scrollView.post {
                    scrollView.scrollTo(0, 0)
                    scrollView.viewTreeObserver.removeOnGlobalLayoutListener(listener)
                }
            }
        })
    }

    override fun onBackPressed() {
        if (mListViewSearch.isVisible){
            mListViewSearch.visibility = View.GONE
            mEditTextSearch.clearFocus()
        } else {
            super.onBackPressed()
        }
    }
}